### ISSUE TITLE
Default Peano opt (~-O2) miscompiles a per-timestep kernel with Buffer state + heavy aie_api vector code; -O1 is correct

### ISSUE THREAD

**Issue by vule20** [Xilinx/mlir-aie#3792]
### What happened?

-----

Default Peano opt (~-O2) miscompiles a per-timestep kernel with Buffer state + heavy aie_api vector code; -O1 is correct

## TL;DR

Same IRON design + same C++ kernel source on NPU2 (AIE2P):

| `compile_flags` | vs pure-numpy golden |
|-----------------|----------------------|
| `-O1` | **MATCH** (bit-exact) |
| `-O2` (IRON/Peano default) | **MISMATCH** (deterministic wrong logits) |

`-O0` crashes the compiler on this kernel. Workaround: pass `compile_flags=["-O1"]` to `ExternalFunction`.

**Runnable repro** (this directory): `python3 run_repro.py`  
Expects: `-O1` MATCH, `-O2` MISMATCH on the same seeded input.

## Environment

- Platform: AMD Ryzen AI Strix Halo, NPU2 (`1022:17f0`)
- mlir-aie IRON (built from source) + XRT 2.21.x
- `llvm-aie` / Peano (wheel or from-source)

## What the design does (minimal shape)

- One Shim + MemTile + CoreTile
- Whole input buffer → MemTile; `consumer_obj_type` meters **per-timestep chunks** to the core (`depth=2`)
- Core loops `range_(T)` calling the same `ExternalFunction` each step
- Recurrence state in an explicit IRON `Buffer` (not a C `static`)
- Kernel: fill small scalar array from state → heavy `aie::accum` / `load_v` / `accumulate` matvec-style reduction → nested scalar loops that re-read the small array → write state back; final step emits 3× int32 logits

A domain-neutral *stub* with the same ObjectFifo/`Buffer`/`range_` wiring but a tiny dummy kernel did **not** reproduce the wrong values. Trigger depends on fuller code / weight-table footprint (repro ships opaque int16 tables). Happy to iterate on a smaller failing case with guidance.

## Ruled out

1. SROA of `static` state — switching to explicit `Buffer` did not fix it  
2. Stale `@iron.jit` cache alone — real, but not sufficient  
3. `range_` vs Python-unrolled calls — no difference  
4. `__attribute__((noinline))` around the vector helper — no difference  

Only changing the Peano optimization level moved the needle.

## Asm / schedule notes (local Peano build)

Exposed-pipeline / no HW stall logic (llvm-aie README). Diffing `-O1` vs `-O2` objects:

- `step` function ~4× larger at `-O2` (heavy unroll)
- After `jl` into the vector helper, `-O1` reloads pointers from the stack; `-O2` trusts live `p6` and soft-pipelines loads
- AIE2P TableGen: `LoadLatency = 5; // FIXME` in `AIE2PSchedule.td`; LDA spill itineraries use `dst=7`; `II_RET` has almost no delay-slot / cross-return modeling

Suspect: scheduler under-NOP’s a dependent use of a restored callee-saved pointer / load result at `-O2`.

## Workaround

```python
ExternalFunction(..., compile_flags=["-O1"])
```

Confirmed on hardware for this design. Does **not** fix an unrelated persistent-core latency issue (`while_true=True` slowdown) — different bug.

## Ask

1. Is this a known Peano/AIE2P scheduling bug (pass between `-O1` and `-O2`)?  
2. Pointers for a TableGen / PostRA fix we could upstream (happy to bisect with a local Peano build).  
3. Any interest in a lit/hardware regression once a fix lands?

## Attachments

- This folder (or `peano_o2_miscompile.zip` from `./make_zip.sh`)
- Optional: `docs/asm_o1_vs_default/` asm dumps from the parent project

[peano_o2_miscompile.zip](https://github.com/user-attachments/files/32543363/peano_o2_miscompile.zip)

### Hardware / device

NPU2

### Version

b710f1b8 (2026-09-22)

### Relevant logs

```shell
=== subprocess -O1 ===                                                                         
REPRO_OPT_LEVEL=1  npu=[-99, 1643, -1896]  golden=[-99, 1643, -1896]  MATCH
                                                                                               
=== subprocess -O2 ===                                                                         
REPRO_OPT_LEVEL=2  npu=[-2059, 3232, -2041]  golden=[-99, 1643, -1896]  MISMATCH
                                               
=== summary ===                                                                                
  -O1: MATCH  npu=[-99, 1643, -1896]  golden=[-99, 1643, -1896]                           
  -O2: MISMATCH  npu=[-2059, 3232, -2041]  golden=[-99, 1643, -1896]
```

**Comment by jgmelber** [comment]
Please file this issue on the llvm-aie (Peano) repository
